### PR TITLE
Fix implementation of scalarProduct

### PR DESCRIPTION
**Pull request by christiankral**
Fix implementation according to [variant 1](https://github.com/modelica/ModelicaStandardLibrary/issues/1260#issuecomment-723048778) of #1260

### CODE REVIEW

File: Complex.mo [hunk 1]
```diff
@@ -94,24 +94,23 @@ operator record Complex "Complex number with overloaded operators"
 </html>"));
     end multiply;
 
-    function scalarProduct "Scalar product c1*c2 of two complex vectors"
+    function scalarProduct "Scalar product of two complex vectors c1 and c2"
       import Complex;
       input Complex c1[:] "Vector of Complex numbers 1";
       input Complex c2[size(c1,1)] "Vector of Complex numbers 2";
-      output Complex c3 "= c1*c2";
+      output Complex c3 "= Scalar product of c1 and c2";
```
Comment (tobolar) [inline Complex.mo:101]: IMO the equal sign "=" can be deleted here (and so the space right behind it).

Comment (HansOlsson) [reply Complex.mo:101]: agreed

**Review by HansOlsson — APPROVED** [review]
Looks good, but I don't know how we should handle conversions of existing code.

File: Complex.mo (hunk 1 above)
Comment (AHaumer) [reply Complex.mo:101]: That's really a minor problem ...

**Review by AHaumer — CHANGES_REQUESTED** [review]
Despite the fact the old implementation was wrong  the bugfix is not backwards compatible: 
The problem is if some user models rely on the old (wrong) implementation.
I'd suggest to:

- move Christian's correct implementation to a new function (no idea for a smart name)
- leave the old implementation unchanged and mark it as obsolete
- include links in both functions' documenation to each other
- state the the old (obsolete) implementation will be deleted, and the the new correct implementation's name could be changed in futute

**Comment by christiankral** [comment]
> Despite the fact the old implementation was wrong the bugfix is not backwards compatible:

A bug fix is not backwards compatible. Otherwise it were not a bug fix. 

So I wonder if -- as a general rule -- all models with severe bugs shall be indicated as obsolete models (in the future). I know we marked some buggy models as obsolete in the past. However, I am not very happy with the idea of marking this model as obsolete, even though the underlying process is very structured and clear. In this case we also "loose" the particular name `scalarProduct` as the new bug fixed model has to have a different name. This leads to more inconsistency in the naming structure of Modelica classes. Additionally the original buggy model has to moved to the ObsoleteModelica package at one point in time, which needs a conversion script. 

One alternative were to indicate a bug fixed model with a bug fix overlay icon in addition to the documentation. In this case the user also notices that a severe bug has been fixed without the need of re-structuring the MSL later on. 

However, if the majority is in favor of indicating `scalarProduct` as obsolete block and creating a new bug fixed block, e.g. `productScalar` I will not oppose. 

I guess we need more discussion here. @HansOlsson @beutlich @dietmarw could you please comment on this issue.

**Comment by dietmarw** [comment]
In my opinion, if it is a bug then it should be fixed in place even if generating a different outcome. Should somebody rely on the buggy behaviour then this would mean the implementation was also flawed and needs fixing anyway. So for me it would be acceptable to "break" such models as long as the bugfix and its impact is clearly documented. 

Introducing a new class for marking bugfixed models seems strange and would just lead to more maintenance work and need to be monitored very carefully.

**Comment by beutlich** [comment]
I agree with @christiankral and @dietmarw here. Bugfixes shall be fixed in place and documented in the revision annotation. I also doubt that anyone used such a function if it was plainly wrong.

**Comment by HansOlsson** [comment]
> I agree with @christiankral and @dietmarw here. Bugfixes shall be fixed in place and documented in the revision annotation. I also doubt that anyone used such a function if it was plainly wrong.

I would hope not, but I have some vague memories of people wanting the wrong behaviour in some cases. 
(And I agree that it may make sense to compute sum(a[i]*b[i] for i) for complex numbers, even though it's not the scalar product.)

**Comment by beutlich** [comment]
What about fixing it in place, moving the buggy implementation to ObsoleteModelica4 and mentioning the revision annotation where to find the old implementation? Conversion is not considered.

**Review by AHaumer — APPROVED** [review]

**Comment by christiankral** [comment]
Re-triggered check

File: Complex.mo [hunk 2]
```diff
@@ -94,24 +94,23 @@ operator record Complex "Complex number with overloaded operators"
...
+                      c3.im + c1[i].re * c2[i].im - c1[i].im * c2[i].re);
       end for;
 
-    annotation(Inline=true, smoothOrder=100, Documentation(info="<html>
-<p>This function returns the scalar product of two given arrays of Complex numbers.</p>
-</html>"));
+    annotation(Inline=true, smoothOrder=100, Documentation(info = "<html><head></head><body><p>This function returns the scalar product of two given vectors of Complex numbers of length <code>n</code>.</p>
+<blockquote><pre>c3 = sum(conj(c1[k]) * c2[k] for k in 1:n)
+</pre></blockquote>
+    </body></html>",
```
Comment (beutlich) [inline Complex.mo:112]:
```suggestion
    annotation(Inline=true, smoothOrder=100, Documentation(info = "<html><p>This function returns the scalar product of two given vectors of Complex numbers of length <code>n</code>.</p>
<blockquote><pre>c3 = sum(conj(c1[k]) * c2[k] for k in 1:n)
</pre></blockquote>
</html>",
```

**Pull request merged**